### PR TITLE
protect internal index in elasticsearch service

### PR DESCRIPTION
**Pull request by stafyniaksacha**
Protect usage of `storageEngine` to manipulate what happen in `internalIndex`

## Breaking changes
* Actions like `write:deleteFromQuery`, `read:search`, `admin:refreshIndex` with `index: %kuzzle` are now discarded with a `BadRequestError`

## Changes
* Added a route `admin:searchSpecifications` to avoid using `read:search`
* Create/Update/Delete actions on `internalEngine` are now send with `refresh=true` to ES
* Rewrite functional tests (`cleanSecurity` & `cleanValidation`)
* Changed default mapping for internal collections (users / profiles / roles / specifications) to match with ES 5.x

### Todo (other PR)
* Remove `allowInternalIndex` from policies (now it's obsolete)
* Add a way to delete at least users (maybe also validation specification, profiles and roles) from a query

### CODE REVIEW

File: lib/services/internalEngine/bootstrap.js [hunk 1]
```diff
@@ -82,8 +82,7 @@ InternalEngineBootstrap.prototype.createProfilesCollection = function internalEn
       policies: {
         properties: {
           _id: {
-            index: 'not_analyzed',
-            type: 'string'
+            type: 'keyword'
```
Comment (dbengsch) [inline lib/services/internalEngine/bootstrap.js]: Not compatible with ES 2

File: lib/services/internalEngine/bootstrap.js [hunk 2]
```diff
@@ -112,12 +111,11 @@ InternalEngineBootstrap.prototype.createUsersCollection = function internalEngin
   return this.engine.updateMapping('users', {
     properties: {
       profileIds: {
-        index: 'not_analyzed',
-        type: 'string'
+        type: 'keyword'
```
Comment (dbengsch) [inline lib/services/internalEngine/bootstrap.js]: Same as above

File: lib/services/internalEngine/bootstrap.js [hunk 3]
```diff
@@ -112,12 +111,11 @@ InternalEngineBootstrap.prototype.createUsersCollection = function internalEngin
...
       },
       password: {
-        index: 'no',
-        type: 'string'
+        index: 'false',
+        type: 'keyword'
```
Comment (dbengsch) [inline lib/services/internalEngine/bootstrap.js]: Same as above

File: lib/services/internalEngine/bootstrap.js [hunk 4]
```diff
@@ -136,12 +134,10 @@ InternalEngineBootstrap.prototype.createValidationCollection = function internal
   return this.engine.updateMapping('validations', {
     properties: {
       index: {
-        index: 'not_analyzed',
-        type: 'string'
+        type: 'keyword'
```
Comment (dbengsch) [inline lib/services/internalEngine/bootstrap.js]: Same as above

File: lib/services/internalEngine/bootstrap.js [hunk 5]
```diff
@@ -136,12 +134,10 @@ InternalEngineBootstrap.prototype.createValidationCollection = function internal
...
       },
       collection: {
-        index: 'not_analyzed',
-        type: 'string'
+        type: 'keyword'
```
Comment (dbengsch) [inline lib/services/internalEngine/bootstrap.js]: Same as above

File: test/services/internalEngine/bootstrap.test.js [hunk 6]
```diff
@@ -219,8 +220,7 @@ describe('services/internalEngine/bootstrap.js', () => {
                   policies: {
                     properties: {
                       _id: {
-                        index: 'not_analyzed',
-                        type: 'string'
+                        type: 'keyword'
```
Comment (dbengsch) [inline test/services/internalEngine/bootstrap.test.js]: Same as above

File: test/services/internalEngine/bootstrap.test.js [hunk 7]
```diff
@@ -277,20 +277,19 @@ describe('services/internalEngine/bootstrap.js', () => {
               .be.calledWithMatch('users', {
                 properties: {
                   profileIds: {
-                    index: 'not_analyzed',
-                    type: 'string'
+                    type: 'keyword'
```
Comment (dbengsch) [inline test/services/internalEngine/bootstrap.test.js]: Same as above

File: test/services/internalEngine/bootstrap.test.js [hunk 8]
```diff
@@ -277,20 +277,19 @@ describe('services/internalEngine/bootstrap.js', () => {
...
                   },
                   password: {
-                    index: 'no',
-                    type: 'string'
+                    index: 'false',
+                    type: 'keyword'
```
Comment (dbengsch) [inline test/services/internalEngine/bootstrap.test.js]: Same as above

File: lib/services/elasticsearch.js [hunk 9]
```diff
@@ -959,13 +969,19 @@ module.exports = ElasticSearch;
...
-function cleanData(requestObject) {
+function cleanData(requestObject, kuzzle) {
   var data = {};
 
   if (requestObject.index !== undefined) {
+    /**
+     * @todo: protect internal index to avoid using this service from requests and core. this let allowInternalIndex policy obsolete
+     */
+    if (requestObject.index === kuzzle.internalEngine.index) {
+      throw new BadRequestError(`Index "${kuzzle.internalEngine.index}" is protected, please use appropriated routes instead`);
```
Comment (benoitvidis) [inline lib/services/elasticsearch.js]: '"Cannot operate on Kuzzle internal index "${kuzzle.internalEngine}"'

File: lib/services/internalEngine/index.js [hunk 10]
```diff
@@ -171,6 +171,7 @@ function InternalEngine (kuzzle) {
   this.create = function internalEngineCreate (type, id, content) {
     var
       request = {
+        refresh: true,
```
Comment (benoitvidis) [inline lib/services/internalEngine/index.js]: I am not sure to agree on this.
If an application has some massive user writes, we do not want to issue a refresh each time.

I don't see any use case where within Kuzzle, we cannot control whether a manual refresh is needed or not.

**Pull request merged**